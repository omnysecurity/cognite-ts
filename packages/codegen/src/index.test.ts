import type { DataModel, ViewDefinition } from '@cognite/sdk';
import { describe, it, expect } from 'vitest';
import { generate } from './';
// import { writeFileSync } from 'fs';

const VIEW_DEFINITIONS = [
	{
		externalId: 'Assessment',
		space: 'sp_test',
		version: 'c902f759da6320',
		createdTime: 1755860256522,
		lastUpdatedTime: 1760431261505,
		writable: true,
		queryable: true,
		usedFor: 'all',
		isGlobal: false,
		properties: {
			status: {
				constraintState: {},
				type: {
					type: 'enum',
					values: {
						Draft: {
							description:
								'A preliminary assessment that is still being prepared or edited and not yet shared beyond the assessor.',
						},
						Provisional: {
							description:
								'An assessment that has been recorded and may be visible, but is subject to change pending review, approval, or additional input.',
						},
						Final: {
							description:
								'A completed assessment that is confirmed, officially recorded, and no longer subject to revision.',
						},
					},
					list: false,
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Assessment',
				},
				containerPropertyIdentifier: 'status',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description:
					'Assessment status describes the stage of completeness and authority of an assessment, indicating whether it is an initial draft, a provisional version awaiting confirmation, or a final record that is officially approved and fixed.\n@name Assessment.status',
				name: 'Assessment.status',
			},
			comment: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Assessment',
				},
				containerPropertyIdentifier: 'comment',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Assessment.comment',
				name: 'Assessment.comment',
			},
		},
		mappedContainers: [
			{
				type: 'container',
				space: 'sp_test',
				externalId: 'Assessment',
			},
		],
		name: 'Assessment',
		description:
			'An Assessment is a textual comment that provides qualitative feedback on a student, professor, subject, or school, describing performance or quality without assigning a formal grade.',
		implements: [],
	},
	{
		externalId: 'Professor',
		space: 'sp_test',
		version: 'a9604ac0c4a94e',
		createdTime: 1755860256522,
		lastUpdatedTime: 1760431261505,
		writable: true,
		queryable: true,
		usedFor: 'node',
		isGlobal: false,
		properties: {
			name: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Professor',
				},
				containerPropertyIdentifier: 'name',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Professor.name',
				name: 'Professor.name',
			},
			school: {
				constraintState: {},
				type: {
					type: 'direct',
					list: false,
					source: {
						type: 'view',
						space: 'sp_test',
						externalId: 'School',
						version: '1cc15f29cb5b7a',
					},
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Professor',
				},
				containerPropertyIdentifier: 'school',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Professor.school',
				name: 'Professor.school',
			},
			colleagues: {
				type: {
					space: 'sp_test',
					externalId: 'hasColleague',
				},
				source: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Professor',
					version: 'a9604ac0c4a94e',
				},
				connectionType: 'multi_edge_connection',
				name: 'Professor.colleagues',
				description: 'Best work buddies\n@name Professor.colleagues',
				direction: 'outwards',
			},
		},
		mappedContainers: [
			{
				type: 'container',
				space: 'sp_test',
				externalId: 'Professor',
			},
		],
		name: 'Professor',
		description:
			'A Professor is an educator responsible for teaching subjects and guiding students within a school.',
		implements: [],
	},
	{
		externalId: 'School',
		space: 'sp_test',
		version: '1cc15f29cb5b7a',
		createdTime: 1755860256522,
		lastUpdatedTime: 1760431261505,
		writable: true,
		queryable: true,
		usedFor: 'node',
		isGlobal: false,
		properties: {
			name: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'School',
				},
				containerPropertyIdentifier: 'name',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name School.name',
				name: 'School.name',
			},
			description: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'School',
				},
				containerPropertyIdentifier: 'description',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name School.description',
				name: 'School.description',
			},
			country: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'School',
				},
				containerPropertyIdentifier: 'country',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				name: 'country',
			},
			students: {
				connectionType: 'multi_reverse_direct_relation',
				source: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Student',
					version: '307de9bcb3ee59',
				},
				through: {
					source: {
						type: 'view',
						space: 'sp_test',
						externalId: 'Student',
						version: '307de9bcb3ee59',
					},
					identifier: 'school',
				},
				targetsList: false,
				name: '*School.students',
				description: '@name *School.students',
			},
		},
		mappedContainers: [
			{
				type: 'container',
				space: 'sp_test',
				externalId: 'School',
			},
		],
		name: 'School',
		description:
			'A School is an organized institution where students are taught by educators in a structured environment to gain knowledge, skills, and cultural or social learning.',
		implements: [],
	},
	{
		externalId: 'Student',
		space: 'sp_test',
		version: '307de9bcb3ee59',
		createdTime: 1755860256522,
		lastUpdatedTime: 1760431261505,
		writable: true,
		queryable: true,
		usedFor: 'node',
		isGlobal: false,
		properties: {
			name: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Student',
				},
				containerPropertyIdentifier: 'name',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Student.name',
				name: 'Student.name',
			},
			birthyear: {
				constraintState: {},
				type: {
					type: 'int32',
					list: false,
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Student',
				},
				containerPropertyIdentifier: 'birthyear',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Student.birthyear',
				name: 'Student.birthyear',
			},
			school: {
				constraintState: {},
				type: {
					type: 'direct',
					list: false,
					source: {
						type: 'view',
						space: 'sp_test',
						externalId: 'School',
						version: '1cc15f29cb5b7a',
					},
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Student',
				},
				containerPropertyIdentifier: 'school',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Student.school',
				name: 'Student.school',
			},
			subjects: {
				type: {
					space: 'sp_test',
					externalId: 'studentEnrolledInClass',
				},
				source: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Subject',
					version: '4d020fa33ecd51',
				},
				connectionType: 'multi_edge_connection',
				name: 'Student.subjects',
				description: '@name Student.subjects',
				edgeSource: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Assessment',
					version: 'c902f759da6320',
				},
				direction: 'outwards',
			},
		},
		mappedContainers: [
			{
				type: 'container',
				space: 'sp_test',
				externalId: 'Student',
			},
		],
		name: 'Student',
		description:
			'A Student is a learner enrolled in a school who attends subjects to gain knowledge and skills.',
		implements: [],
	},
	{
		externalId: 'Subject',
		space: 'sp_test',
		version: '4d020fa33ecd51',
		createdTime: 1755860256522,
		lastUpdatedTime: 1760431261505,
		writable: true,
		queryable: true,
		usedFor: 'node',
		isGlobal: false,
		properties: {
			school: {
				constraintState: {},
				type: {
					type: 'direct',
					list: false,
					source: {
						type: 'view',
						space: 'sp_test',
						externalId: 'School',
						version: '1cc15f29cb5b7a',
					},
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Subject',
				},
				containerPropertyIdentifier: 'school',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Subject.school',
				name: 'Subject.school',
			},
			professor: {
				constraintState: {},
				type: {
					type: 'direct',
					list: false,
					source: {
						type: 'view',
						space: 'sp_test',
						externalId: 'Professor',
						version: 'a9604ac0c4a94e',
					},
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Subject',
				},
				containerPropertyIdentifier: 'professor',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Subject.professor',
				name: 'Subject.professor',
			},
			name: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Subject',
				},
				containerPropertyIdentifier: 'name',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Subject.name',
				name: 'Subject.name',
			},
			description: {
				constraintState: {},
				type: {
					type: 'text',
					list: false,
					collation: 'ucs_basic',
				},
				container: {
					type: 'container',
					space: 'sp_test',
					externalId: 'Subject',
				},
				containerPropertyIdentifier: 'description',
				immutable: false,
				nullable: true,
				autoIncrement: false,
				description: '@name Subject.description',
				name: 'Subject.description',
			},
			students: {
				type: {
					space: 'sp_test',
					externalId: 'studentEnrolledInClass',
				},
				source: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Student',
					version: '307de9bcb3ee59',
				},
				connectionType: 'multi_edge_connection',
				name: 'Subject.students',
				description: '@name Subject.students',
				edgeSource: {
					type: 'view',
					space: 'sp_test',
					externalId: 'Assessment',
					version: 'c902f759da6320',
				},
				direction: 'inwards',
			},
		},
		mappedContainers: [
			{
				type: 'container',
				space: 'sp_test',
				externalId: 'Subject',
			},
		],
		name: 'Subject',
		description:
			"A Subject is a defined area of study within a school's curriculum that is taught by a professor.",
		implements: [],
	},
] as unknown as ViewDefinition[];

describe('generate', () => {
	it('generates code', () => {
		const output = generate({
			dataModel: {
				space: 'TEST',
				externalId: 'TEST',
				version: '1',
			} as DataModel,
			views: VIEW_DEFINITIONS,
		});

		// record new snapshot using pnpm test -- -u
		expect(output).toMatchSnapshot();
		// writeFileSync('.generated.ts', output, { encoding: 'utf-8' });
	});
});
