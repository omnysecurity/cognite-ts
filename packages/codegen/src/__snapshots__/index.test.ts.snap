// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > generates code 1`] = `
{
  "fileContent": "/*
 * This file was generated by @omnysec/cognite-codegen.
 * It is based on the data model 'TEST' version '1.'
 */
export type __Schema = {
    sp_test__Assessment__c902f759da6320: sp_test__Assessment__c902f759da6320;
    sp_test__Professor__a9604ac0c4a94e: sp_test__Professor__a9604ac0c4a94e;
    sp_test__School__1cc15f29cb5b7a: sp_test__School__1cc15f29cb5b7a;
    sp_test__Student__307de9bcb3ee59: sp_test__Student__307de9bcb3ee59;
    sp_test__Subject__4d020fa33ecd51: sp_test__Subject__4d020fa33ecd51;
};
export type DirectReference<T extends __Schema[keyof __Schema] | unknown = unknown> = Omit<T, keyof T> & {
    space: string;
    externalId: string;
};
/**
 * An Assessment is a textual comment that provides qualitative feedback on a student, professor, subject, or school, describing performance or quality without assigning a formal grade. */
export type sp_test__Assessment__c902f759da6320 = {
    status?: "Draft" | "Provisional" | "Final";
    comment?: string;
};
/**
 * A Professor is an educator responsible for teaching subjects and guiding students within a school. */
export type sp_test__Professor__a9604ac0c4a94e = {
    name?: string;
    school?: DirectReference<sp_test__School__1cc15f29cb5b7a>;
};
/**
 * A School is an organized institution where students are taught by educators in a structured environment to gain knowledge, skills, and cultural or social learning. */
export type sp_test__School__1cc15f29cb5b7a = {
    name?: string;
    description?: string;
    country?: string;
};
/**
 * A Student is a learner enrolled in a school who attends subjects to gain knowledge and skills. */
export type sp_test__Student__307de9bcb3ee59 = {
    name?: string;
    birthyear?: number;
    school?: DirectReference<sp_test__School__1cc15f29cb5b7a>;
};
/**
 * A Subject is a defined area of study within a school's curriculum that is taught by a professor. */
export type sp_test__Subject__4d020fa33ecd51 = {
    school?: DirectReference<sp_test__School__1cc15f29cb5b7a>;
    professor?: DirectReference<sp_test__Professor__a9604ac0c4a94e>;
    name?: string;
    description?: string;
};
export const __VIEWS = {
    sp_test__Assessment__c902f759da6320: ["status", "comment"],
    sp_test__Professor__a9604ac0c4a94e: ["name", "school", "colleagues"],
    sp_test__School__1cc15f29cb5b7a: ["name", "description", "country", "students"],
    sp_test__Student__307de9bcb3ee59: ["name", "birthyear", "school", "subjects"],
    sp_test__Subject__4d020fa33ecd51: ["school", "professor", "name", "description", "students"]
} as const;
export const __DATA_MODEL = {
    space: "TEST",
    externalId: "TEST",
    version: "1"
} as const;


const _VIEW_DEFINITIONS = [
  {
    "externalId": "Assessment",
    "space": "sp_test",
    "version": "c902f759da6320",
    "createdTime": 1755860256522,
    "lastUpdatedTime": 1760431261505,
    "writable": true,
    "queryable": true,
    "usedFor": "all",
    "isGlobal": false,
    "properties": {
      "status": {
        "constraintState": {},
        "type": {
          "type": "enum",
          "values": {
            "Draft": {
              "description": "A preliminary assessment that is still being prepared or edited and not yet shared beyond the assessor."
            },
            "Provisional": {
              "description": "An assessment that has been recorded and may be visible, but is subject to change pending review, approval, or additional input."
            },
            "Final": {
              "description": "A completed assessment that is confirmed, officially recorded, and no longer subject to revision."
            }
          },
          "list": false
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Assessment"
        },
        "containerPropertyIdentifier": "status",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "Assessment status describes the stage of completeness and authority of an assessment, indicating whether it is an initial draft, a provisional version awaiting confirmation, or a final record that is officially approved and fixed.\\n@name Assessment.status",
        "name": "Assessment.status"
      },
      "comment": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Assessment"
        },
        "containerPropertyIdentifier": "comment",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Assessment.comment",
        "name": "Assessment.comment"
      }
    },
    "mappedContainers": [
      {
        "type": "container",
        "space": "sp_test",
        "externalId": "Assessment"
      }
    ],
    "name": "Assessment",
    "description": "An Assessment is a textual comment that provides qualitative feedback on a student, professor, subject, or school, describing performance or quality without assigning a formal grade.",
    "implements": []
  },
  {
    "externalId": "Professor",
    "space": "sp_test",
    "version": "a9604ac0c4a94e",
    "createdTime": 1755860256522,
    "lastUpdatedTime": 1760431261505,
    "writable": true,
    "queryable": true,
    "usedFor": "node",
    "isGlobal": false,
    "properties": {
      "name": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Professor"
        },
        "containerPropertyIdentifier": "name",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Professor.name",
        "name": "Professor.name"
      },
      "school": {
        "constraintState": {},
        "type": {
          "type": "direct",
          "list": false,
          "source": {
            "type": "view",
            "space": "sp_test",
            "externalId": "School",
            "version": "1cc15f29cb5b7a"
          }
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Professor"
        },
        "containerPropertyIdentifier": "school",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Professor.school",
        "name": "Professor.school"
      },
      "colleagues": {
        "type": {
          "space": "sp_test",
          "externalId": "hasColleague"
        },
        "source": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Professor",
          "version": "a9604ac0c4a94e"
        },
        "connectionType": "multi_edge_connection",
        "name": "Professor.colleagues",
        "description": "Best work buddies\\n@name Professor.colleagues",
        "direction": "outwards"
      }
    },
    "mappedContainers": [
      {
        "type": "container",
        "space": "sp_test",
        "externalId": "Professor"
      }
    ],
    "name": "Professor",
    "description": "A Professor is an educator responsible for teaching subjects and guiding students within a school.",
    "implements": []
  },
  {
    "externalId": "School",
    "space": "sp_test",
    "version": "1cc15f29cb5b7a",
    "createdTime": 1755860256522,
    "lastUpdatedTime": 1760431261505,
    "writable": true,
    "queryable": true,
    "usedFor": "node",
    "isGlobal": false,
    "properties": {
      "name": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "School"
        },
        "containerPropertyIdentifier": "name",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name School.name",
        "name": "School.name"
      },
      "description": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "School"
        },
        "containerPropertyIdentifier": "description",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name School.description",
        "name": "School.description"
      },
      "country": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "School"
        },
        "containerPropertyIdentifier": "country",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "name": "country"
      },
      "students": {
        "connectionType": "multi_reverse_direct_relation",
        "source": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Student",
          "version": "307de9bcb3ee59"
        },
        "through": {
          "source": {
            "type": "view",
            "space": "sp_test",
            "externalId": "Student",
            "version": "307de9bcb3ee59"
          },
          "identifier": "school"
        },
        "targetsList": false,
        "name": "*School.students",
        "description": "@name *School.students"
      }
    },
    "mappedContainers": [
      {
        "type": "container",
        "space": "sp_test",
        "externalId": "School"
      }
    ],
    "name": "School",
    "description": "A School is an organized institution where students are taught by educators in a structured environment to gain knowledge, skills, and cultural or social learning.",
    "implements": []
  },
  {
    "externalId": "Student",
    "space": "sp_test",
    "version": "307de9bcb3ee59",
    "createdTime": 1755860256522,
    "lastUpdatedTime": 1760431261505,
    "writable": true,
    "queryable": true,
    "usedFor": "node",
    "isGlobal": false,
    "properties": {
      "name": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Student"
        },
        "containerPropertyIdentifier": "name",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Student.name",
        "name": "Student.name"
      },
      "birthyear": {
        "constraintState": {},
        "type": {
          "type": "int32",
          "list": false
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Student"
        },
        "containerPropertyIdentifier": "birthyear",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Student.birthyear",
        "name": "Student.birthyear"
      },
      "school": {
        "constraintState": {},
        "type": {
          "type": "direct",
          "list": false,
          "source": {
            "type": "view",
            "space": "sp_test",
            "externalId": "School",
            "version": "1cc15f29cb5b7a"
          }
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Student"
        },
        "containerPropertyIdentifier": "school",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Student.school",
        "name": "Student.school"
      },
      "subjects": {
        "type": {
          "space": "sp_test",
          "externalId": "studentEnrolledInClass"
        },
        "source": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Subject",
          "version": "4d020fa33ecd51"
        },
        "connectionType": "multi_edge_connection",
        "name": "Student.subjects",
        "description": "@name Student.subjects",
        "edgeSource": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Assessment",
          "version": "c902f759da6320"
        },
        "direction": "outwards"
      }
    },
    "mappedContainers": [
      {
        "type": "container",
        "space": "sp_test",
        "externalId": "Student"
      }
    ],
    "name": "Student",
    "description": "A Student is a learner enrolled in a school who attends subjects to gain knowledge and skills.",
    "implements": []
  },
  {
    "externalId": "Subject",
    "space": "sp_test",
    "version": "4d020fa33ecd51",
    "createdTime": 1755860256522,
    "lastUpdatedTime": 1760431261505,
    "writable": true,
    "queryable": true,
    "usedFor": "node",
    "isGlobal": false,
    "properties": {
      "school": {
        "constraintState": {},
        "type": {
          "type": "direct",
          "list": false,
          "source": {
            "type": "view",
            "space": "sp_test",
            "externalId": "School",
            "version": "1cc15f29cb5b7a"
          }
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Subject"
        },
        "containerPropertyIdentifier": "school",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Subject.school",
        "name": "Subject.school"
      },
      "professor": {
        "constraintState": {},
        "type": {
          "type": "direct",
          "list": false,
          "source": {
            "type": "view",
            "space": "sp_test",
            "externalId": "Professor",
            "version": "a9604ac0c4a94e"
          }
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Subject"
        },
        "containerPropertyIdentifier": "professor",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Subject.professor",
        "name": "Subject.professor"
      },
      "name": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Subject"
        },
        "containerPropertyIdentifier": "name",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Subject.name",
        "name": "Subject.name"
      },
      "description": {
        "constraintState": {},
        "type": {
          "type": "text",
          "list": false,
          "collation": "ucs_basic"
        },
        "container": {
          "type": "container",
          "space": "sp_test",
          "externalId": "Subject"
        },
        "containerPropertyIdentifier": "description",
        "immutable": false,
        "nullable": true,
        "autoIncrement": false,
        "description": "@name Subject.description",
        "name": "Subject.description"
      },
      "students": {
        "type": {
          "space": "sp_test",
          "externalId": "studentEnrolledInClass"
        },
        "source": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Student",
          "version": "307de9bcb3ee59"
        },
        "connectionType": "multi_edge_connection",
        "name": "Subject.students",
        "description": "@name Subject.students",
        "edgeSource": {
          "type": "view",
          "space": "sp_test",
          "externalId": "Assessment",
          "version": "c902f759da6320"
        },
        "direction": "inwards"
      }
    },
    "mappedContainers": [
      {
        "type": "container",
        "space": "sp_test",
        "externalId": "Subject"
      }
    ],
    "name": "Subject",
    "description": "A Subject is a defined area of study within a school's curriculum that is taught by a professor.",
    "implements": []
  }
] as const;

type _Expand<T> = T extends infer U ? { [K in keyof U]: U[K] } : never;
type _Mutable<T> = _Expand<{ -readonly [K in keyof T]: _Mutable<T[K]> }>;

type _ViewReferenceCandidate<TSchema> = keyof TSchema extends infer K
	? K extends \`\${infer S}__\${infer E}__\${infer V}\` // full
		? \`\${S}__\${E}__\${V}\` | \`\${S}__\${E}\` | E
		: K extends \`\${infer S}__\${infer E}\` // versioned
			? \`\${S}__\${E}\` | E
			: K // simple
	: never;
type _CollectMatches<
	TSchema,
	TRef extends _ViewReferenceCandidate<TSchema>,
> = keyof TSchema extends infer K
	? K extends keyof TSchema
		? K extends \`\${infer S}__\${infer E}__\${infer V}\`
			? TRef extends \`\${S}__\${E}__\${V}\` | \`\${S}__\${E}\` | E
				? TSchema[K]
				: never
			: K extends \`\${infer S}__\${infer E}\`
				? TRef extends \`\${S}__\${E}\` | E
					? TSchema[K]
					: never
				: TRef extends K
					? TSchema[K]
					: never
		: never
	: never;
type _IsUnion<T, U = T> = T extends any
	? [U] extends [T]
		? false
		: true
	: false;
type _IsSingleValue<T> = [T] extends [never]
	? unknown
	: _IsUnion<T> extends true
		? unknown
		: T;
type ResolveView<
	TSchema,
	TRef extends _ViewReferenceCandidate<TSchema>,
> = _IsSingleValue<_CollectMatches<TSchema, TRef>>;
type _UnambiguousViewReference<TSchema> =
	Extract<_ViewReferenceCandidate<TSchema>, string> extends infer TRef extends
		string
		? TRef extends _ViewReferenceCandidate<TSchema>
			? unknown extends ResolveView<TSchema, TRef>
				? never
				: TRef
			: never
		: never;

export type Schema = {
	[K in _UnambiguousViewReference<__Schema>]: _Expand<ResolveView<__Schema, K>>;
};
export const VIEW_DEFINITIONS = _VIEW_DEFINITIONS as _Mutable<
	typeof _VIEW_DEFINITIONS
>;	
",
  "fileName": "TEST@1.ts",
}
`;
